hydromt/components/table.py

```diff
@@ -0,0 +1,133 @@
+        """Model tables."""
+        if self._data is None:
+            self._initialize_tables()
+
+        assert self._data is not None
+        return self._data
+
+    def _initialize_tables(self, skip_read=False) -> None:
+        """Initialize the model tables."""
+        if self._data is None:
+            self._data = dict()
+            if self._root.is_reading_mode() and not skip_read:
+                self.read()
+
+    @hydromt_step
+    def write(self, fn: str = DEFAULT_TABLE_FILENAME, **kwargs) -> None:
+        """Write tables at <root>/tables."""
+        self._root._assert_write_mode()
+        if len(self.data) > 0:
+            self._model.logger.info("Writing table files.")
+            local_kwargs = {"index": False, "header": True, "sep": ","}
+            local_kwargs.update(**kwargs)
+            for name in self.data:
+                fn_out = join(self._root.path, fn.format(name=name))
+                os.makedirs(dirname(fn_out), exist_ok=True)
+                self.data[name].to_csv(fn_out, **local_kwargs)
+        else:
+            self._model.logger.debug("No tables found, skip writing.")
+
+    @hydromt_step
+    def read(self, fn: str = DEFAULT_TABLE_FILENAME, **kwargs) -> None:
+        """Read table files at <root>/tables and parse to dict of dataframes."""
+        self._root._assert_read_mode()
+        self._initialize_tables(skip_read=True)
+        self._model.logger.info("Reading model table files.")
+        fns = glob.glob(join(self._root.path, fn.format(name="*")))
+        if len(fns) > 0:
+            for fn in fns:
+                name = basename(fn).split(".")[0]
+                tbl = pd.read_csv(fn, **kwargs)
+                self.set(tbl, name=name)
+
+    def set(
+        self,
+        tables: Union[
+            Union[pd.DataFrame, pd.Series], Dict[str, Union[pd.DataFrame, pd.Series]]
+        ],
+        name: Optional[str] = None,
+    ) -> None:
+        """Add (a) table(s) <pandas.DataFrame> to model.
+
+        Parameters
+        ----------
+        tables : pandas.DataFrame, pandas.Series or dict
+            Table(s) to add to model.
+            Multiple tables can be added at once by passing a dict of tables.
+        name : str, optional
+            Name of table, by default None. Required when tables is not a dict.
+        """
+        self._initialize_tables()
+        assert self._data is not None
+        if not isinstance(tables, dict):
+            if name is None:
+                raise ValueError("name required when tables is not a dict")
+            else:
+                tables_to_add: Dict[str, Union[pd.DataFrame, pd.Series]] = {
+                    name: tables
+                }
+        else:
+            tables_to_add: Dict[str, Union[pd.DataFrame, pd.Series]] = tables
+
+        for df_name, df in tables_to_add.items():
+            if not (isinstance(df, pd.DataFrame) or isinstance(df, pd.Series)):
```

Comment:
To be honest, I find all these type checking a tad heavy... And are they really needed, you check higher up on dict and later you check then on pd.DataFrame or pd.Series so this all seems redundant to me. If not I really do need my type checking course ;)

```suggestion
        if not isinstance(tables, dict):
            if name is None:
                raise ValueError("name required when tables is not a dict")
            else:
                tables_to_add = {name: tables}
        else:
            tables_to_add = tables

        for df_name, df in tables_to_add.items():
            if not (isinstance(df, pd.DataFrame) or isinstance(df, pd.Series)):
```

Reply:
Some of these checks are not for us but for the type checker. Because python is such a free and dynamic language it can sometimes not make deductions that are easy to see for us. I moved the type definition to the start so we only have to do it once, so hopefully that will make it a bit easier for you.